WordPress/src/main/java/org/wordpress/android/ui/posts/RemotePreviewLogicHelper.kt

```diff
@@ -46,41 +46,54 @@ class RemotePreviewLogicHelper @Inject constructor(
         post: PostModel,
         helperFunctions: RemotePreviewHelperFunctions
     ): PreviewLogicOperationResult {
-        if (!site.isUsingWpComRestApi) {
-            activityLauncherWrapper.showActionableEmptyView(
-                    activity,
-                    WPWebViewUsageCategory.REMOTE_PREVIEW_NOT_AVAILABLE,
-                    post.title
-            )
-            return PreviewLogicOperationResult.PREVIEW_NOT_AVAILABLE
-        } else if (!networkUtilsWrapper.isNetworkAvailable()) {
+        if (!networkUtilsWrapper.isNetworkAvailable()) {
             activityLauncherWrapper.showActionableEmptyView(
                     activity,
                     WPWebViewUsageCategory.REMOTE_PREVIEW_NO_NETWORK,
                     post.title
             )
             return PreviewLogicOperationResult.NETWORK_NOT_AVAILABLE
-        } else {
-            if (helperFunctions.notifyUploadInProgress(post)) {
-                return PreviewLogicOperationResult.MEDIA_UPLOAD_IN_PROGRESS
-            }
+        }
 
-            val updatedPost = helperFunctions.updatePostIfNeeded() ?: post
+        // If a media upload is currently in progress, we can't upload / auto-save or preview until it's finished.
+        if (helperFunctions.notifyUploadInProgress(post)) {
+            return PreviewLogicOperationResult.MEDIA_UPLOAD_IN_PROGRESS
+        }
 
-            if (shouldUpload(updatedPost)) {
+        // Update the post object (copy title/content from the editor to the post object) when it's needed
+        // (eg. during and editing session)
+        val updatedPost = helperFunctions.updatePostIfNeeded() ?: post
+
+        when {
```

Comment:
⭐️ I love the refactoring here. I have a suggestion to possibly improve this more. Since we already have a few `return PreviewLogicOperationResult` in the cases, we could fully commit to this and use `return when { }`:

```kt
// Note: Removed the other parts
return when {
    shouldUpload(updatedPost) -> {
	if (!postUtilsWrapper.isPublishable(updatedPost)) {
	    return PreviewLogicOperationResult.CANNOT_SAVE_EMPTY_DRAFT
	}

	PreviewLogicOperationResult.GENERATING_PREVIEW
    }
    shouldRemoteAutoSave(updatedPost) -> {
	if (!site.isUsingWpComRestApi) {
	    return PreviewLogicOperationResult.PREVIEW_NOT_AVAILABLE
	}

	if (!postUtilsWrapper.isPublishable(updatedPost)) {
	    return PreviewLogicOperationResult.CANNOT_REMOTE_AUTO_SAVE_EMPTY_POST
	}

	PreviewLogicOperationResult.GENERATING_PREVIEW
    }
    else -> {
	PreviewLogicOperationResult.OPENING_PREVIEW
    }
}
```

For me, this is a bit more easier to follow since we don't need to look at the final `return PreviewLogicOperationResult.GENERATING_PREVIEW` at the bottom of the method.

Reply:
Good point, updated in d1ab054